### PR TITLE
chore: extend JNI to get strings

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

**Comment by eddyxu** [comment]
@LuQQiu any suggestion about how to test this in rust?

File: java/lance-jni/src/ffi.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            let jstr = JString::from(elem);
+            let val = self.get_string(&jstr)?;
+            results.push(val.to_str()?.to_string())
+        }
+        Ok(results)
+    }
+
+    fn get_strings_opt(&mut self, obj: &JObject) -> Result<Option<Vec<String>>> {
+        let is_empty = self.call_method(obj, "java/util/Optional/isEmpty", "()Z", &[])?;
+        if !is_empty.z()? {
```
Comment (LuQQiu) [inline java/lance-jni/src/ffi.rs:44]: Seems logically inverted?

File: java/lance-jni/src/ffi.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        while let Some(elem) = iter.next(self)? {
+            let jstr = JString::from(elem);
+            let val = self.get_string(&jstr)?;
+            results.push(val.to_str()?.to_string())
+        }
+        Ok(results)
+    }
+
+    fn get_strings_opt(&mut self, obj: &JObject) -> Result<Option<Vec<String>>> {
+        let is_empty = self.call_method(obj, "java/util/Optional/isEmpty", "()Z", &[])?;
```
Comment (LuQQiu) [inline java/lance-jni/src/ffi.rs:43]: isEmpty is jdk 11 +, isPresent can be used with jdk 8+, FYI if need to connect with legacy java ecosystem

File: java/lance-jni/src/ffi.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            let inner =
+                self.call_method(obj, "java/util/Optional/get", "()Ljava/util/List;", &[])?;
+            let inner_obj = inner.l()?;
+            Ok(Some(self.get_strings(&inner_obj)?))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {}
```
Comment (LuQQiu) [inline java/lance-jni/src/ffi.rs:56]: Not required for this PR.
JNI tests may be helpful to ensure correctness when we add more complicated JNI methods.
The biggest overhead of JNI related tests is that it requires an actual JVM
```
        let jvm = jni::JavaVM::new(jvm_args).unwrap();
        let env = jvm.attach_current_thread().unwrap();
```

**Review by LuQQiu — CHANGES_REQUESTED** [review]
Thanks for adding the helpful methods, left some comments

File: java/lance-jni/src/ffi.rs (hunk 2 above)
Comment (eddyxu) [reply java/lance-jni/src/ffi.rs:43]: Got it. We dont plan to support JDK 8.

File: java/lance-jni/src/ffi.rs (hunk 3 above)
Comment (eddyxu) [reply java/lance-jni/src/ffi.rs]: 🆗

File: java/lance-jni/src/ffi.rs (hunk 1 above)
Comment (eddyxu) [reply java/lance-jni/src/ffi.rs]: Fixed @LuQQiu , could you take another look?

**Review by LuQQiu — APPROVED** [review]
LGTM! Thanks!

**Review by beinan — APPROVED** [review]

**Pull request merged**